presto-native-execution/presto_cpp/main/common/StatsReporterImpl.h

```diff
@@ -0,0 +1,173 @@
+class StatsReporterImpl : public facebook::velox::BaseStatsReporter {
+ public:
+  StatsReporterImpl(
+      const std::string cluster = "",
+      const std::string worker = "") {
+    if (cluster.empty()) {
+      auto nodeConfig = facebook::presto::NodeConfig::instance();
+      cluster_ = nodeConfig->nodeEnvironment();
+    } else {
+      cluster_ = cluster;
+    }
+    char* hostName = std::getenv("HOSTNAME");
+    workerPod_ = !hostName ? worker : hostName;
+  }
+
+  /// Register a stat of the given stat type.
+  /// @param key The key to identify the stat.
+  /// @param statType How the stat is aggregated.
+  void registerMetricExportType(
+      const char* key,
+      facebook::velox::StatType statType) const override;
+
+  void registerMetricExportType(
+      folly::StringPiece key,
+      facebook::velox::StatType statType) const override;
+
+  void registerHistogramMetricExportType(
+      const char* /*key*/,
+      int64_t /* bucketWidth */,
+      int64_t /* min */,
+      int64_t /* max */,
+      const std::vector<int32_t>& /* pcts */) const override {}
+
+  void registerHistogramMetricExportType(
+      folly::StringPiece /* key */,
+      int64_t /* bucketWidth */,
+      int64_t /* min */,
+      int64_t /* max */,
+      const std::vector<int32_t>& /* pcts */) const override {}
+
+  void addMetricValue(const std::string& key, size_t value = 1) const override;
+
+  void addMetricValue(const char* key, size_t value = 1) const override;
+
+  void addMetricValue(folly::StringPiece key, size_t value = 1) const override;
+
+  void addHistogramMetricValue(const std::string& key, size_t value)
+      const override {}
+
+  void addHistogramMetricValue(const char* key, size_t value) const override {}
+
+  void addHistogramMetricValue(folly::StringPiece key, size_t value)
+      const override {}
+
+  const facebook::velox::StatType getRegisteredStatType(
+      const std::string& metricName) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    return registeredStats_[metricName];
+  }
+
+  /*
+   * Serializes the metrics collected so far in the format suitable for
+   * back filling Prometheus server.
+   *
+   * Given a metric name and a set of labels, time series are frequently
+   * identified using this notation:
+   *
+   *       <metric name>{<label name>=<label value>, ...}
+   *
+   * For example, a time series with the metric name num_tasks_aborted
+   * and the labels cluster="<cluster_id>" and worker="worker-id"
+   * could be written like this:
+   *  # HELP num_tasks_aborted
+   *  # TYPE num_tasks_aborted gauge*
+   *  num_tasks_aborted{cluster="<cluster_id>", worker="worker-id"} value
+   * timestamp
+   *
+   * Above info is from:
+   * https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+   */
+  const std::string getMetrics(const MetricsSerializer& serializer);
+
+ private:
+  /// Mapping of registered stats key to StatType.
+  mutable std::unordered_map<std::string, facebook::velox::StatType>
```

Comment:
Use mutable folly::ConcurrentHashMap

Reply:
we normally use `folly::Synchronized<>` to do this. Lmk revisit this after finishing the RFC. I wanted to unblock any testing you wanted to do, so focussed on getting a working version. I need to clean up the tests as well.